packages/extension/src/shared/transactions/index.ts

```diff
@@ -129,3 +133,295 @@ function buildGetTransactionsFn(metadataTransactions: Transaction[]) {
+      return signedChainedTxResultToReviewTransactionResult(chainedParams[index], signedChainedTxResult, networkId)
+    })
+  } catch (error) {
+    console.log("Error building chained transaction", error)
+    throw error
+  }
+}
+
+export async function tryBuildTransactions(
+  nodeUrl: string,
+  tokensWithBalance: BaseTokenWithBalance[],
+  selectedAccount: WalletAccount,
+  allAccounts: WalletAccount[],
+  transactionParams: TransactionParams,
+  useLedger: boolean
+): Promise<ReviewTransactionResult[]> {
+  const builder = TransactionBuilder.from(nodeUrl)
+  try {
+    const transaction: ReviewTransactionResult = await buildTransaction(builder, selectedAccount, transactionParams)
+    return [transaction]
+  } catch (error) {
+    console.log("Error building transaction", error)
+    const missingBalances = await checkBalances(tokensWithBalance, transactionParams)
+    const isDAppTransaction = transactionParams.type === 'EXECUTE_SCRIPT' || transactionParams.type === 'DEPLOY_CONTRACT'
+
+    if (missingBalances !== undefined) {
+      if (isDAppTransaction && !useLedger) {
+        const restOfAccounts = allAccounts.filter((account) => account.address !== selectedAccount.address)
+        const nodeProvider = new NodeProvider(nodeUrl)
+        const account = await accountWithEnoughBalance(missingBalances, restOfAccounts, nodeProvider)
+
+        if (account !== undefined) {
+          const tokens = Array.from(missingBalances.entries())
+            .filter(([tokenId]) => tokenId !== ALPH_TOKEN_ID)
+            .map(([tokenId, amount]) => ({
+              id: tokenId,
+              amount: amount.toString()
+            }))
+          const attoAlphAmount = missingBalances.get(ALPH_TOKEN_ID)?.toString() || DUST_AMOUNT * BigInt(tokens.length)
+          const transferTransactionParams: SignTransferChainedTxParams = {
+            type: 'Transfer',
+            signerAddress: account.address,
+            signerKeyType: account.signer.keyType,
+            destinations: [{
+              address: selectedAccount.address,
+              attoAlphAmount,
+              tokens
+            }]
+          }
+
+          const dappTransactionParams = transactionParamsToSignChainedTxParams(transactionParams)
+          const chainedTxParams = [transferTransactionParams, dappTransactionParams]
+          const chainedTxResults = await builder.buildChainedTx(
+            chainedTxParams,
+            [account.signer.publicKey, selectedAccount.signer.publicKey]
+          )
+
+          const reviewTxResults = chainedTxParams.map((params, index) => {
+            return signedChainedTxResultToReviewTransactionResult(params, chainedTxResults[index], account.networkId)
+          })
+
+          return reviewTxResults
+        }
+      }
+
+      const [firstMissingTokenId, firstMissingAmount] = missingBalances.entries().next().value;
+      const tokenSymbol = firstMissingTokenId === ALPH_TOKEN_ID ? 'ALPH' : firstMissingTokenId;
```

Comment:
Is it possible to get the token symbol based on the token id?

Reply:
Addressed now, thanks :+1: